### PR TITLE
Refactor lib to use plonk 

### LINKED ISSUE #5: Port Poseidon to Plonk
We no longer require curve25519 nor bulletproofs. 

Instead we will use Bls12_381 and plonk to work with gadgets and ZK.
So, we need to update the dependencies and refactor the code where needed.

### PR DESCRIPTION
**Pull request by CPerezz**
## Adds
- travis basic configfile
- Optimization on the Sponge hash fn to skip useless addition if the width of the messages is the appropiate one to do it. (WIDTH < HADES_WIDTH -1).
- CHANGELOG & CONTRIBUTING files.
## Changes
- Sponge hash refactor with gadget and scalar versions.

## Removes
- Old tests which no longer apply.

Closes #5

I bumped to 0.2, so if this gets merged I'll tag it but probably not release it since it's not finnished. @ZER0 let me know WDYT.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (ZER0) [inline CONTRIBUTING.md]: I guess it should be "Poseidon"

File: CONTRIBUTING.md
Comment (ZER0) [inline CONTRIBUTING.md:31]: I would also add somewhere a reference to [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/)

File: Cargo.toml
Comment (ZER0) [inline Cargo.toml]: Add yourself to the list of authors, as well as Victor.

File: Cargo.toml
Comment (ZER0) [inline Cargo.toml]: Which version of Hades are we targeting?

File: src/sponge/pad.rs
Comment (ZER0) [inline src/sponge/pad.rs]: for consistency maybe we could call it `eom_value` at this point?

File: src/sponge/pad.rs
Comment (ZER0) [inline src/sponge/pad.rs:3]: Maybe we could be more descriptive about what it does, at least how the `pad_value` and `eom` are used.

File: src/sponge/pad.rs
Comment (ZER0) [inline src/sponge/pad.rs]: I would also rename in the tests `zero` and `one` here as `pad_value` and `eom_value` when are used for such purpose, it would be clearer.

File: src/sponge/pad.rs
Comment (ZER0) [inline src/sponge/pad.rs]: Shall we remove it, then?

File: src/sponge/sponge.rs
Comment (ZER0) [inline src/sponge/sponge.rs:1]: It's a bit unclear to me this comment.

File: src/sponge/sponge.rs
Comment (ZER0) [inline src/sponge/sponge.rs]: For readability, I would do something like:
```rust
// Filling the vector returned by `pad` with zero
let pad_value = Scalar::zero();
// one will identify the end of messages
let eom_value = Scalar::one();

let mut words = pad(messages, WIDTH, pad_value, eom_value);
```

Or something like that.

File: src/sponge/sponge.rs
Comment (ZER0) [inline src/sponge/sponge.rs:46]: Ah ah ah, I like the fact that you ALREADY did what I suggested in the review for the gadget part. :) So, yeah, be consistent, you can add `padder` and `eom` also for the scalar's one.

File: src/sponge/sponge.rs
Comment (ZER0) [inline src/sponge/sponge.rs:73]: At this point you know better than me: since you always pass the same `composer`, do we need to create a new strategy all the time as well? Shouldn't we create it just once after the composer?

File: src/sponge/sponge.rs
Comment (ZER0) [inline src/sponge/sponge.rs]: What's the comment line stands for

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM, I would just change few nits and I've a couple of questions, but it's good overall

File: Cargo.toml
Comment (CPerezz) [reply Cargo.toml]: Shouldn't we wait for #11 to land to then make the bump?

File: src/sponge/pad.rs
Comment (CPerezz) [reply src/sponge/pad.rs:3]: Not targetting to have a good documentation atm.

I just put this so I don't get errors due to the flag `#[deny(missing_docs)]`.
I was leaving docs for the end once #11 gets merged.

File: src/sponge/sponge.rs
Comment (CPerezz) [reply src/sponge/sponge.rs:1]: Not targetting to have a good documentation atm.

I just put this so I don't get errors due to the flag #[deny(missing_docs)].
I was leaving docs for the end once #11 gets merged.

File: src/sponge/sponge.rs
Comment (CPerezz) [reply src/sponge/sponge.rs]: Some testing debuggin I was doing. Mb

File: src/sponge/sponge.rs
Comment (CPerezz) [reply src/sponge/sponge.rs:73]: Since it is mutably borrowed, I cannot create the strategy at the beggining of the function since then the `&mut Composer` would be already borrowed by the strategy and I will not be able to call it's functions or attributes.

File: Cargo.toml
Comment (CPerezz) [reply Cargo.toml]: I'm picking "master" since there's no tag or version for the newes't implementation which is using PLONK as constraint system & proof system.

I think that you were going to add the tag once the history was re-written. But I checked and it hasn't been created.

File: src/sponge/pad.rs
Comment (CPerezz) [reply src/sponge/pad.rs]: Yepp

File: src/sponge/sponge.rs
Comment (CPerezz) [reply src/sponge/sponge.rs]: Yep, agree!

File: src/sponge/sponge.rs
Comment (CPerezz) [reply src/sponge/sponge.rs:46]: Done avobe!

File: CONTRIBUTING.md
Comment (ZER0) [inline CONTRIBUTING.md]: For consistency, I would actually wrote:
`See [How to Write a Git Commit Message]( https://chris.beams.io/posts/git-commit/)`

File: src/sponge/pad.rs
Comment (ZER0) [inline src/sponge/pad.rs]: Since you used `padder` and `eom` as identifiers later on, I think we should be consistent and use them in the tests as well. Or viceversa. Whatever you like more.

File: Cargo.toml
Comment (ZER0) [inline Cargo.toml]: As discussed, point to `0.5.0` already. Even if it fails the build on travis it will pass once we add the tag.

**Review by ZER0 — COMMENTED** [review]
LGTM, with the few nits addressed.

**Review by ZER0 — APPROVED** [review]

**Comment by CPerezz** [comment]
@ZER0 feel free to merge as soon as it passes! :)

**Pull request merged**
